src/core.rs

```diff
@@ -671,30 +906,60 @@ const HYPHEN_PENALTY: i32 = 25;
 /// by Knuth and Plass. The implementation here is based on [Python
 /// code by David
 /// Eppstein](https://github.com/jfinkels/PADS/blob/master/pads/wrap.py).
-pub fn wrap_optimal_fit<'a, T: Fragment, F: Fn(usize) -> usize>(
-    fragments: &'a [T],
-    line_widths: F,
-) -> Vec<&'a [T]> {
-    let mut widths = Vec::with_capacity(fragments.len() + 1);
-    let mut width = 0;
-    widths.push(width);
-    for fragment in fragments {
-        width += fragment.width() + fragment.whitespace_width();
-        widths.push(width);
+pub fn wrap_optimal_fit<F, W>(
+    fragments: F,
+    line_widths: W,
+) -> WrapOptimalFit<<F as IntoIterator>::Item>
```

Comment:
Is there a reason not to let the functions return `impl Iterator` instead of concrete structs? That should slim down the API, just like I did in #201.

Reply:
For both of the iterators I have a `.terminate_eol()` function, which I think is very useful - I use both (terminating and non-terminating) variants in different parts of the library. It could be added as an extra parameter to the wrapping functions of course, but it makes it harder to call the functions and it is not obvious what the `true` in something like `wrap_optimal_fit(fragments, widths, true)` would mean - on the other hand `wrap_optimal_fit(fragments, widths).terminate_eol()` is very clear.

Also, `WrapOptimalFit::into_vec` is useful for efficiency, and it couldn't exist with `impl Iterator`.

Reply:
I haven't yet figured out what this `bool` is for :see_no_evil:  but I haven't look super careful at the code yet.

Reply:
I looked more at the code and I think I see it now: instead of yielding full lines, your iterators yield _once per `Fragment`_. The `bool` here tells the caller if the `Fragment` you just saw is at the end of a line.

This sounds expensive to me since the caller needs to "touch" every fragment. The existing code allows the caller in `wrap_first_fit` to simply deal with a full line at a time.

Reply:
> Also, `WrapOptimalFit::into_vec` is useful for efficiency, and it couldn't exist with `impl Iterator`.

Would it be possible to implement `Into<Vec>` instead?